src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +52,55 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
-
-    def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
-      if (reserveThread()) {
-        wire(new ForkJoinWorkerThread(fjp) with BlockContext {
-          // We have to decrement the current thread count when the thread exits
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
-                  }
-
-                true
-              }
-              override def isReleasable = isdone
-            })
-            result
+    private[this] final def allocateNormalThread(): Boolean =
```

Comment:
I am not able to review this logic carefully. I understood 'scala.concurrent.blocking' clause now has got only effect when current number of already blocked threads is < scala.concurrent.context.maxExtraThreads. Is that right? 
I achieved the same logic by placing a semaphore before invoking scala.concurrent.blocking. I think applying a semaphore would help to make this code easier, but up to you..

Reply:
@avkonst No, what the code does is that it only spawns compensating threads up to maxExtraThreads.